File: backend/app/api/session_keys.py
"""
Session Keys API endpoints.

Endpoints:
- /api/v1/zkdefi/session_keys/grant - Generate session key request
- /api/v1/zkdefi/session_keys/revoke - Revoke session key
- /api/v1/zkdefi/session_keys/list - List active sessions
- /api/v1/zkdefi/session_keys/validate - Validate session key
"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.services.session_key_service import get_session_service

router = APIRouter()


# ==================== Request Models ====================

class GrantSessionRequest(BaseModel):
    """Request to grant a new session key."""
    owner_address: str
    session_key_address: str
    max_position: int
    allowed_protocols: list[str]  # ["pools", "ekubo", "jediswap"]
    duration_hours: int = 24


class ConfirmGrantRequest(BaseModel):
    """Confirm session grant on-chain."""
    session_id: str
    tx_hash: str


class RevokeSessionRequest(BaseModel):
    """Request to revoke a session key."""
    session_id: str
    owner_address: str


class ConfirmRevokeRequest(BaseModel):
    """Confirm session revoke on-chain."""
    session_id: str
    tx_hash: str


class ValidateSessionRequest(BaseModel):
    """Validate an action under a session."""
    session_id: str
    protocol_id: int
    amount: int


# ==================== Endpoints ====================

@router.post("/grant")
async def grant_session(data: GrantSessionRequest):
    """
    Generate a session key grant request.
    
    Returns calldata for the wallet to sign and submit on-chain.
    """
    try:
        service = get_session_service()
        result = await service.generate_session_request(
            owner_address=data.owner_address,
            session_key_address=data.session_key_address,
            max_position=data.max_position,
            allowed_protocols=data.allowed_protocols,
            duration_hours=data.duration_hours
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/grant/confirm")
async def confirm_grant(data: ConfirmGrantRequest):
    """
    Confirm that a session was granted on-chain.
    """
    try:
        service = get_session_service()
        result = await service.confirm_session_grant(
            session_id=data.session_id,
            tx_hash=data.tx_hash
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/revoke")
async def revoke_session(data: RevokeSessionRequest):
    """
    Generate a session revocation request.
    
    Returns calldata for the wallet to sign and submit on-chain.
    """
    try:
        service = get_session_service()
        result = await service.revoke_session(
            session_id=data.session_id,
            owner_address=data.owner_address
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/revoke/confirm")
async def confirm_revoke(data: ConfirmRevokeRequest):
    """
    Confirm that a session was revoked on-chain.
    """
    try:
        service = get_session_service()
        result = await service.confirm_session_revoke(
            session_id=data.session_id,
            tx_hash=data.tx_hash
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/list/{owner_address}")
async def list_sessions(owner_address: str):
    """
    List all sessions for a user.
    """
    try:
        service = get_session_service()
        sessions = await service.list_user_sessions(owner_address=owner_address)
        return {
            "owner_address": owner_address,
            "sessions": sessions,
            "count": len(sessions),
            "active_count": len([s for s in sessions if s["is_active"]])
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/validate")
async def validate_session(data: ValidateSessionRequest):
    """
    Validate if an action is allowed under a session.
    """
    try:
        service = get_session_service()
        result = await service.validate_session(
            session_id=data.session_id,
            protocol_id=data.protocol_id,
            amount=data.amount
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend.preview.deploy/src/lib/sessionKeys.ts
/**
 * Session Keys Library
 * 
 * Utilities for managing session keys with Starknet account abstraction.
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8003";

export interface SessionKeyConfig {
  sessionKeyAddress: string;
  maxPosition: number;
  allowedProtocols: string[];
  durationHours: number;
}

export interface Session {
  sessionId: string;
  sessionKey: string;
  maxPosition: number;
  allowedProtocols: string[];
  durationHours: number;
  createdAt: string;
  expiresAt: string;
  isActive: boolean;
  isExpired: boolean;
}

/**
 * Generate a session key grant request
 */
export async function generateSessionRequest(
  ownerAddress: string,
  config: SessionKeyConfig
): Promise<{
  sessionId: string;
  calldata: Record<string, any>;
  contractAddress: string;
  entrypoint: string;
}> {
  const response = await fetch(`${API_BASE}/api/v1/zkdefi/session_keys/grant`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      owner_address: ownerAddress,
      session_key_address: config.sessionKeyAddress,
      max_position: config.maxPosition,
      allowed_protocols: config.allowedProtocols,
      duration_hours: config.durationHours,
    }),
  });

  if (!response.ok) {
    throw new Error("Failed to generate session request");
  }

  const data = await response.json();
  return {
    sessionId: data.session_id,
    calldata: data.calldata,
    contractAddress: data.contract_address,
    entrypoint: data.entrypoint,
  };
}

/**
 * Confirm session grant on-chain
 */
export async function confirmSessionGrant(
  sessionId: string,
  txHash: string
): Promise<void> {
  const response = await fetch(`${API_BASE}/api/v1/zkdefi/session_keys/grant/confirm`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      session_id: sessionId,
      tx_hash: txHash,
    }),
  });

  if (!response.ok) {
    throw new Error("Failed to confirm session grant");
  }
}

/**
 * Revoke a session key
 */
export async function revokeSession(
  sessionId: string,
  ownerAddress: string
): Promise<{
  calldata: Record<string, any>;
  contractAddress: string;
  entrypoint: string;
}> {
  const response = await fetch(`${API_BASE}/api/v1/zkdefi/session_keys/revoke`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      session_id: sessionId,
      owner_address: ownerAddress,
    }),
  });

  if (!response.ok) {
    throw new Error("Failed to revoke session");
  }

  const data = await response.json();
  return {
    calldata: data.calldata,
    contractAddress: data.contract_address,
    entrypoint: data.entrypoint,
  };
}

/**
 * Get all sessions for a user
 */
export async function getUserSessions(ownerAddress: string): Promise<Session[]> {
  const response = await fetch(
    `${API_BASE}/api/v1/zkdefi/session_keys/list/${ownerAddress}`
  );

  if (!response.ok) {
    throw new Error("Failed to fetch sessions");
  }

  const data = await response.json();
  return (data.sessions || []).map((s: any) => ({
    sessionId: s.session_id,
    sessionKey: s.session_key,
    maxPosition: s.max_position,
    allowedProtocols: s.allowed_protocols,
    durationHours: s.duration_hours,
    createdAt: s.created_at,
    expiresAt: s.expires_at,
    isActive: s.is_active,
    isExpired: s.is_expired,
  }));
}

/**
 * Validate if an action is allowed under a session
 */
export async function validateSession(
  sessionId: string,
  protocolId: number,
  amount: number
): Promise<{
  isValid: boolean;
  reason?: string;
  remainingTimeSeconds?: number;
}> {
  const response = await fetch(`${API_BASE}/api/v1/zkdefi/session_keys/validate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      session_id: sessionId,
      protocol_id: protocolId,
      amount: amount,
    }),
  });

  if (!response.ok) {
    throw new Error("Failed to validate session");
  }

  const data = await response.json();
  return {
    isValid: data.is_valid,
    reason: data.reason,
    remainingTimeSeconds: data.remaining_time_seconds,
  };
}

/**
 * Calculate time remaining for a session
 */
export function getTimeRemaining(expiresAt: string): {
  hours: number;
  minutes: number;
  isExpired: boolean;
} {
  const expires = new Date(expiresAt);
  const now = new Date();
  const diff = expires.getTime() - now.getTime();

  if (diff <= 0) {
    return { hours: 0, minutes: 0, isExpired: true };
  }

  const hours = Math.floor(diff / (1000 * 60 * 60));
  const minutes = Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60));

  return { hours, minutes, isExpired: false };
}

/**
 * Format time remaining as string
 */
export function formatTimeRemaining(expiresAt: string): string {
  const { hours, minutes, isExpired } = getTimeRemaining(expiresAt);

  if (isExpired) return "Expired";
  if (hours > 0) return `${hours}h ${minutes}m`;
  return `${minutes}m`;
}

/**
 * Convert protocol names to bitmap
 */
export function protocolsToBitmap(protocols: string[]): number {
  const protocolMap: Record<string, number> = {
    pools: 1,
    ekubo: 2,
    jediswap: 4,
  };

  return protocols.reduce((bitmap, protocol) => {
    return bitmap | (protocolMap[protocol.toLowerCase()] || 0);
  }, 0);
}

/**
 * Convert bitmap to protocol names
 */
export function bitmapToProtocols(bitmap: number): string[] {
  const protocols: string[] = [];
  if (bitmap & 1) protocols.push("pools");
  if (bitmap & 2) protocols.push("ekubo");
  if (bitmap & 4) protocols.push("jediswap");
  return protocols;
}
